src/databricks/labs/ucx/sequencing/sequencing.py

```diff
@@ -0,0 +1,171 @@
+
+
+@dataclass
+class MigrationStep:
+    step_id: int
+    step_number: int
+    object_type: str
+    object_id: str
+    object_name: str
+    object_owner: str
+    required_step_ids: list[int] = field(default_factory=list)
+
+
+@dataclass
+class MigrationNode:
+    last_node_id = 0
+    node_id: int
+    object_type: str
+    object_id: str
+    object_name: str
+    object_owner: str
+    required_steps: list[MigrationNode] = field(default_factory=list)
+
+    def generate_steps(self) -> tuple[MigrationStep, Iterable[MigrationStep]]:
+        # traverse the nodes using a depth-first algorithm
+        # ultimate leaves have a step number of 1
+        # use highest required step number + 1 for this step
+        highest_step_number = 0
+        required_step_ids: list[int] = []
+        all_generated_steps: list[Iterable[MigrationStep]] = []
+        for required_step in self.required_steps:
+            step, generated_steps = required_step.generate_steps()
+            highest_step_number = max(highest_step_number, step.step_number)
+            required_step_ids.append(step.step_id)
+            all_generated_steps.append(generated_steps)
+            all_generated_steps.append([step])
+        this_step = MigrationStep(
+            step_id=self.node_id,
+            step_number=highest_step_number + 1,
+            object_type=self.object_type,
+            object_id=self.object_id,
+            object_name=self.object_name,
+            object_owner=self.object_owner,
+            required_step_ids=required_step_ids,
+        )
+        return this_step, itertools.chain(*all_generated_steps)
+
+    def find(self, object_type: str, object_id: str) -> MigrationNode | None:
+        if object_type == self.object_type and object_id == self.object_id:
+            return self
+        for step in self.required_steps:
+            found = step.find(object_type, object_id)
+            if found:
+                return found
+        return None
+
+
+class MigrationSequencer:
+
+    def __init__(self, ws: WorkspaceClient):
+        self._ws = ws
+        self._root = MigrationNode(
+            node_id=0, object_type="ROOT", object_id="ROOT", object_name="ROOT", object_owner="NONE"
+        )
+
+    def register_workflow_task(self, task: jobs.Task, job: jobs.Job, _graph: DependencyGraph) -> MigrationNode:
+        task_id = f"{job.job_id}/{task.task_key}"
+        task_node = self._find_node(object_type="TASK", object_id=task_id)
+        if task_node:
+            return task_node
+        job_node = self.register_workflow_job(job)
+        MigrationNode.last_node_id += 1
+        task_node = MigrationNode(
+            node_id=MigrationNode.last_node_id,
+            object_type="TASK",
+            object_id=task_id,
+            object_name=task.task_key,
+            object_owner=job_node.object_owner,  # no task owner so use job one
+        )
+        job_node.required_steps.append(task_node)
```

Comment:
this algorithm is a bit convoluted and hard to follow. please rewrite with the adjacency map and do a topological sorting with kahn's algorithm in this class. e.g.

```python
self._adjacency = collections.defaultdict(set)

...

# have a central _nodes field to have all the nodes addressable by (TYPE, ID) tuple
self._nodes[('TASK', task_id)] = MigrationNode(
            node_id=self._last_node_id,
            object_type="TASK",
            object_id=task_id,
            object_name=task.task_key,
            object_owner=job_node.object_owner,
        )
self._adjacency[('JOB', job_id)].append(('TASK', task_id))
```

... and actual toposort is pretty straightforward with Kahn's algo:

```python
indegrees = collections.defaultdict(int)
for src, dep_set in self._adjacency.items():
  indegrees[src] = len(dep_set) # count incoming dependencies for a node

queue, toposorted, sequence_num = [], [], 0
for src, incoming in indegrees.items():
  if incoming > 0: continue
  queue.append(src) # start with zero-dependencies nodes

while queue:
  curr = queue.popleft()
  toposorted.append(replace(self._nodes[curr], sequence_num=sequence_num))
  sequence_num += 1
  for dep in self._adjacency[curr]:
    indegrees[dep] -= 1
    if indegrees[dep] == 0:
      queue.append(dep)

return toposorted
```

i find it confusing with all those `_deduplicate_steps` and `_find_node` methods.

Reply:
The problem with Kahn is it only works for DAGs, which we can't guarantee. We have duplicates and recursive loops. I'm trying it out but I suspect it will break on solacc.

Reply:
Consider the following code:

file a.py:
```
class A:
 
    def return_a_bee(self):
        from b import B
        return B()

```
file b.py:

```
from a import A

class B(A):
     pass
```

the above will return the following dependency graph fragment:

    T
    / \
  A   B
 /       \
B       A

When building the dependency graph we detect the recursive cycle (and we break it).
In the above there is no 0-dependency node, so not sure Kahn will work...